### PR TITLE
[Iterators] Implement new ZipOp.

### PR DESCRIPTION
**Pull request by ingomueller-net**

### CODE REVIEW

File: experimental/iterators/include/iterators/Dialect/Iterators/IR/IteratorsOps.td [hunk 1]
```diff
@@ -432,6 +432,46 @@ def Iterators_ValueToStreamOp : Iterators_Op<"value_to_stream",
...
+                    .cast<::mlir::TupleType>().getTypes()}]],
+              "result stream must consist of tuples whose element types match "
+              "the element types of the input streams">,
+     DeclareOpInterfaceMethods<OpAsmOpInterface, ["getAsmResultNames"]>]> {
+  let summary = "Zips several streams into a stream of tuples.";
+  let description = [{
+    Reads one or more streams in lock step and produces a stream of tuples where
+    each struct constists of the elements of all input streams at the same
+    position of the streams.
+
```
Comment (nicolasvasilache) [inline experimental/iterators/include/iterators/Dialect/Iterators/IR/IteratorsOps.td:452]: You want to describe the termination condition.

For instance LLVM also as zip_equal that asserts all streams have the same number of elements.

File: experimental/iterators/lib/Conversion/IteratorsToLLVM/IteratorsToLLVM.cpp [hunk 2]
```diff
@@ -1291,6 +1291,172 @@ static Value buildStateCreation(ValueToStreamOp op,
...
+    SmallVector<Type> nextResultTypes = {upstreamStateType, i1,
+                                         inputElementType};
+    SymbolRefAttr nextFunc = upstreamInfo.nextFunc;
+    auto nextCall =
+        b.create<func::CallOp>(nextFunc, nextResultTypes, initialUpstreamState);
+    Value upstreamHasNext = nextCall->getResult(1);
+    Value upstreamElement = nextCall->getResult(2);
+
+    // Combine hasNext value of the call with previous ones.
+    hasNext = b.create<arith::AndIOp>(hasNext, upstreamHasNext);
```
Comment (nicolasvasilache) [inline experimental/iterators/lib/Conversion/IteratorsToLLVM/IteratorsToLLVM.cpp:1389]: We need more description for the termination state.
You could also envision having attributes so your op works with zip_min/zip_equal and other types of semantics, asserting properly.

File: experimental/iterators/include/iterators/Dialect/Iterators/IR/IteratorsOps.td (hunk 1 above)
Comment (ingomueller-net) [reply experimental/iterators/include/iterators/Dialect/Iterators/IR/IteratorsOps.td:446]: OK, done.

File: experimental/iterators/lib/Conversion/IteratorsToLLVM/IteratorsToLLVM.cpp (hunk 2 above)
Comment (ingomueller-net) [reply experimental/iterators/lib/Conversion/IteratorsToLLVM/IteratorsToLLVM.cpp:1422]: OK, documented in TD file and added as TODO.

**Review by ingomueller-net — COMMENTED** [review]
Thanks for the review! Addressed, merging...

**Pull request merged**